staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go

```diff
@@ -0,0 +1,145 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	namespace = "apiserver"
+	subsystem = "flowcontrol"
+)
+
+const (
+	priorityLevel = "priorityLevel"
+	flowSchema    = "flowSchema"
+)
+
+var (
+	queueLengthBuckets            = []float64{0, 10, 25, 50, 100, 250, 500, 1000}
+	requestDurationSecondsBuckets = []float64{0, 0.005, 0.02, 0.05, 0.1, 0.2, 0.5, 1, 2, 5, 10, 30}
+)
+
+func init() {
+	prometheus.MustRegister(apiserver_rejected_requests)
+	prometheus.MustRegister(apiserver_current_inqueue_requests)
+	prometheus.MustRegister(apiserver_request_queue_length)
+	prometheus.MustRegister(apiserver_request_concurrency_limit)
+	prometheus.MustRegister(apiserver_current_executing_requests)
+	prometheus.MustRegister(apiserver_requests_waiting_seconds)
+	prometheus.MustRegister(apiserver_requests_execution_seconds)
+}
+
+var (
+	apiserver_rejected_requests = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "apiserver_rejected_requests",
+			Help:      "Number of rejected requests by request-management system",
+		},
+		[]string{priorityLevel, "reason"},
+	)
+	apiserver_current_inqueue_requests = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "apiserver_current_inqueue_requests",
+			Help:      "Number of requests currently pending in the queue by request-management system",
+		},
+		[]string{priorityLevel},
+	)
+	apiserver_request_queue_length = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "apiserver_request_queue_length",
+			Help:      "Length of queue in the request-management system",
+			Buckets:   queueLengthBuckets,
+		},
+		[]string{priorityLevel},
+	)
+	apiserver_request_concurrency_limit = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "apiserver_request_concurrency_limit",
+			Help:      "Limit of shared concurrency in the request-management system",
+		},
+		[]string{priorityLevel},
+	)
+	apiserver_current_executing_requests = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "apiserver_current_executing_requests",
+			Help:      "Number of requests currently executing in the request-management system",
+		},
+		[]string{priorityLevel},
+	)
+	apiserver_requests_waiting_seconds = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "apiserver_requests_waiting_inqueue_seconds",
+			Help:      "Time of requests pending in the queue",
```

Comment:
The wording here is a bit awkward.  I suggest something more like "Length of time a request spent waiting in its queue".